internal/config/variables/variables.go

```diff
@@ -0,0 +1,531 @@
+// after the runner has decoded the variables defined in the waypoint.hcl.
+// All values are set as protobuf strings, with the expectation that later
+// evaluation will convert them to their defined types.
+func SetJobInputValues(vars map[string]string, files []string) ([]*pb.Variable, hcl.Diagnostics) {
+	var diags hcl.Diagnostics
+	ret := []*pb.Variable{}
+
+	// The order here is important, as the order in which values are evaluated
+	// dictate their precedence. We therefore evalute these three sources in order
+	// of env, file, cli.
+
+	// process env values ("env" source)
+	{
+		env := os.Environ()
+		for _, raw := range env {
+			if !strings.HasPrefix(raw, varEnvPrefix) {
+				continue
+			}
+			raw = raw[len(varEnvPrefix):] // trim the prefix
+
+			eq := strings.Index(raw, "=")
+			if eq == -1 {
+				// Seems invalid, so we'll ignore it.
+				continue
+			}
+
+			name := raw[:eq]
+			rawVal := raw[eq+1:]
+
+			ret = append(ret, &pb.Variable{
+				Name:   name,
+				Value:  &pb.Variable_Str{Str: rawVal},
+				Source: &pb.Variable_Env{},
+			})
+		}
+	}
+
+	// process -var-file args ("file" source)
+	for _, file := range files {
+		if file != "" {
+			pbv, diags := parseFileValues(file, sourceFile)
+			if diags.HasErrors() {
+				return nil, diags
+			}
+			ret = append(ret, pbv...)
+		}
+	}
+
+	// process -var args ("cli" source)
+	for name, val := range vars {
+		ret = append(ret, &pb.Variable{
+			Name:   name,
+			Value:  &pb.Variable_Str{Str: val},
+			Source: &pb.Variable_Cli{},
+		})
+	}
+	return ret, diags
+}
+
+// EvalInputValues evaluates the provided variable values and validates their
+// types per the type declared in the waypoint.hcl for that variable name.
+// The order in which values are evaluated corresponds to their precedence, with
+// higher precedence values overwriting lower precedence values.
+func EvalInputValues(
+	pbvars []*pb.Variable,
+	vs map[string]*Variable,
+) (InputValues, hcl.Diagnostics) {
+	var diags hcl.Diagnostics
+	iv := InputValues{}
+
+	for v, def := range vs {
+		iv[v] = def.Default
+	}
+
+	for _, pbv := range pbvars {
+		variable, found := vs[pbv.Name]
+		if !found {
+			// TODO krantzinator: this should be a DiagWarning rather than a DiagErr
+			// What to do with a warning diag type?
+			diags = append(diags, &hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Undefined variable",
+				Detail: fmt.Sprintf("A %q variable value was set, "+
+					"but was not found in known variables. To declare variable "+
+					"%q, place a variable definition block in your waypoint.hcl file.",
+					pbv.Name, pbv.Name),
+			})
+			continue
+		}
+
+		var source string
+		switch pbv.Source.(type) {
+		case *pb.Variable_Cli:
+			source = sourceCLI
+		case *pb.Variable_File_:
+			source = sourceFile
+		case *pb.Variable_Env:
+			source = sourceEnv
+		case *pb.Variable_Vcs:
+			source = sourceVCS
+		case *pb.Variable_Server:
+			source = sourceServer
+		}
```

Comment:
I recommend a default case here that logs or errors, otherwise `source` is empty string and who knows how that cacades.

Reply:
I'm going to log it. If there's an error here, it's because our code isn't setting a source correctly somewhere, and I don't want to break a user's ability to use this if we mess up a code path.